### ISSUE TITLE
Improve splitting algorithm

### ISSUE THREAD

**Issue by TimUntersberger** [TimUntersberger/nog#99]
>In what sequence did you create the tiles?

For something like this:

![Screenshot from 2020-08-03 08-55-10](https://user-images.githubusercontent.com/1421719/89184967-805abf00-d567-11ea-8af0-78a84426dca9.png)

(Assuming "open window" is just like... opening a new application that automatically gets managed)

1) Open window (A)
2) Open window with vertical split (B)
3) Switch to horizontal split and open window (C)
4) Switch back to vertical split and open window (D)

I don't really have a use case for this, it feels like a contrived situation I came across playing around with this but I can't expand B individually width-wise without affecting C. 

Also, I did run into this afterward while closing windows I had opened and resized:
>thread 'main' panicked at 'attempt to divide by zero', src/tile_grid.rs:491:30

But I'm having trouble reproducing that.

_Originally posted by @ramirezmike in https://github.com/TimUntersberger/wwm/issues/93#issuecomment-668017480_

**Comment by ramirezmike** [comment]
I haven't looked at the code for splitting it yet but I did think about how it behaves a bit more and it's like when you have tiles A, B and C made and try to make a tile to the right of C.. it behaves like this:

There exists Column1 and Column2. 

Column1 
- contains TileA

Column2
- contains Row1 which contains TileB
- contains Row2 which contains TileC

creating TileD to the right of TileC behaves like creating 

Column3 
- contains Row3 that is empty
- contains Row4 with TileD

 **next** to Column2 instead of _altering_ Column2 to be this:

Column2
- contains Row1 which contains TileB
- contains Row2
  - contains Column3 which contains TileC
  - contains Column4 which contains TileD

Does that make sense or am I way off? It's like a Tile should be able to live in a column or row but if you introduce a row under it then you have to also place the focused tile in a row and the same behavior but with columns. I can create a visual if needed.

**Comment by TimUntersberger** [comment]
# How it currently works

**Grid state before**
| A | B |
|---|---|
| A | C |

focus is on tile C
the split direction is set to `Vertical`
you open a new window

**Grid state after**
| A | B |   |
|---|---|---|
| A | C | D |

# How it will work

I will be using the same row/column approach I used for resizing. This will simplify the implementation and reduce the amount of edge cases.

You can still set the split direction for each tile, but the new location is no longer relative to the current tile, but to the current row/column.

**Grid state before**
| A | B |
|---|---|
| A | C |

focus is on tile C
the split direction is set to `Vertical`
you open a new window

**Grid state after**
| A | B | D |
|---|---|---|
| A | C | D |

**Comment by ramirezmike** [comment]
In the proposed way, is it possible to get to a state like this?

A | B | B
------------ | ------------- | -------------
A | C | D

**Comment by TimUntersberger** [comment]
I am noticing so many flaws in the way I am currently doing splits, that I think I will have to completely rewrite this. 😢

Short answer: No

**Comment by ramirezmike** [comment]
Again, I don't have a use case for that, it's just an example scenario that maybe someone might try. If it's any consolation, I'm already finding the existing implementation incredibly useful for my workflow and am using it almost 100% of the time I'm working on windows machines.

**Comment by ramirezmike** [comment]
So, I've been a bit busy recently but found some time to think about how to handle splitting and made a little proof of concept of an approach that seems to work well. Basically it's to represent everything in a graph. I used [petgraph](https://docs.rs/crate/petgraph/0.5.1) and it seems to work pretty well in my prototype.

The prototype I made is a terminal app and the groupings of letters represent program windows; I figure the graph handles the relationships and sizes and the "content" of the nodes would just be window handles and maybe some other meta properties. 

Here are a couple gifs of it handling different scenarios, but keep in mind that this is essentially running on a grid of 126 x 15 characters. Also, apologies for the "flashing," the way it's recorded kinda introduces some drawing artifacts like that.


Basic splitting, moving a tile and removing a tile:
![01](https://user-images.githubusercontent.com/1421719/92305917-9fc77c00-ef59-11ea-88e4-3a982df9059f.gif)


This is the scenario that we talked about above:
![abcd](https://user-images.githubusercontent.com/1421719/92305936-bec60e00-ef59-11ea-8b37-553027988a91.gif)


And here's an example of how flexible it can be. I don't really think anyone would setup anything like this but it can support it.
![examples](https://user-images.githubusercontent.com/1421719/92306073-8b37b380-ef5a-11ea-92bd-e7e193f85e41.gif)


Here's moving tiles:
![move](https://user-images.githubusercontent.com/1421719/92306109-be7a4280-ef5a-11ea-9dfc-a5fb6f260f0c.gif)


Removing tiles: 
![remove](https://user-images.githubusercontent.com/1421719/92306122-c89c4100-ef5a-11ea-9c9e-7d2fcc9f6d2e.gif)


Resizing tiles:
![resize](https://user-images.githubusercontent.com/1421719/92306124-cf2ab880-ef5a-11ea-8881-3ba802fd918a.gif)


And here's an example of that rotating thing I mentioned some time ago. This takes all columns and turns them into rows and vice versa:
![rotate](https://user-images.githubusercontent.com/1421719/92306163-ec5f8700-ef5a-11ea-9977-ae5f472cd938.gif)


The way I wrote it, I can resize the terminal and then redraw and everything adjusts to the new "resolution" as well, so that could be a way of handling redrawing workspaces across monitors/unplugging monitors. I might push a branch to my fork later today with some initial stuff, but.. yeah, let me know what you think. If you dig it, my plan would be to get it to functional parity with what's already in nog and then add additional features like the rotate.

**Comment by TimUntersberger** [comment]
@ramirezmike holy that looks really good. 

Maybe being able to switch to a terminal based text renderer while developing nog could be useful. We could seperate the grid logic from the winapi calls and then just use a "plug and play" system to change the renderer. (I can open a separate issue for this)

> So, I've been a bit busy recently but found some time to think about how to handle splitting and made a little proof of concept of an approach that seems to work well. Basically it's to represent everything in a graph. I used petgraph and it seems to work pretty well in my prototype.

I am really excited to see your solution using petgraph!

**Comment by ramirezmike** [comment]
Ok, sorry this is going to be a bunch of text.

So I've been working on this a bit and am *near* feature parity. I'm able to use it on my work machine at this point and am just playing around with it. I'm just going through at this point and making sure the focus behavior is the same and tweaking how sizes are handled a bit and then I'll need to add padding/margin which I don't think will be too bad.

(this is it running on a pretty empty windows 10 machine so I didn't have a lot of neat apps to position)

![Screenshot from 2020-09-24 21-35-10](https://user-images.githubusercontent.com/1421719/94217395-e7ce2480-feaf-11ea-8fd1-c754cfc3f7cb.png)

First, going off your suggestion, I implemented the "renderer" by making it take a resolution and then it looks at the graph and returns a list of the tiles along with their x/y/width/height. Then, I have it just call for the screen's resolution for the "graph renderer" and the resolution of the log file for a "text renderer" and figured I'd make a config setting for displaying it in the log. 

So, this is the log of the above:

![Screenshot from 2020-09-24 21-35-59](https://user-images.githubusercontent.com/1421719/94218312-38468180-feb2-11ea-9957-4643beff8e81.png)

I initially tried to have it display the name of the tile but instead just went with letters in each tile followed by a legend that maps out what each lettered tile represents.

SO, while playing with this I found that I really wanted three ways of moving: swapping, moving into and moving out of. Swapping is pretty straightforward. If you have a set of tiles in a column you can swap up/down. If you have a set of tiles in a row, you can swap left/right.

I also wanted to take a tile and tell nog to move it into a column next to it or into the tile next to it to create a row, but at the same time I wanted the ability to do the opposite and getting it to predict what the user wants in some cases is impossible. So, I made some new config controls:
```

    bind "W" focus("Up");
    bind "A" focus("Left");
    bind "S" focus("Down");
    bind "D" focus("Right");

    bind "H" swap("Left");
    bind "J" swap("Down");
    bind "K" swap("Up");
    bind "L" swap("Right");

    bind "Alt+H" move_in("Left");
    bind "Alt+J" move_in("Down");
    bind "Alt+K" move_in("Up");
    bind "Alt+L" move_in("Right");

    bind "Alt+Shift+H" move_out("Left");
    bind "Alt+Shift+J" move_out("Down");
    bind "Alt+Shift+K" move_out("Up");
    bind "Alt+Shift+L" move_out("Right");

```

At a high level, the graph holds nodes that are either "column" nodes, "row" nodes or tile nodes. Column/row behave the same way in that all of their children should display as either columns or rows respectively. Tiles are just tiles. 

So, the above example would be expressed in the graph as 

![Screenshot from 2020-09-24 22-31-31](https://user-images.githubusercontent.com/1421719/94219839-b8bab180-feb5-11ea-8631-3f62eac98578.png)

Again, a column node just indicates that all its children should be represented as columns. And the same with rows but with rows.

So, onto the controls... given the following setup...

![Screenshot from 2020-09-24 21-44-44](https://user-images.githubusercontent.com/1421719/94218958-a475b500-feb3-11ea-81ac-cb351b9b6148.png)


With the middle tile focused, swapping to the left behaves as expected:

![Screenshot from 2020-09-24 21-45-06](https://user-images.githubusercontent.com/1421719/94218983-b22b3a80-feb3-11ea-86ee-b54472dc5d6a.png)

If instead, I did a "move_in(right)" it would take the focused tile and combine it with the tile to the right of it under a row node. 

![Screenshot from 2020-09-24 21-45-20](https://user-images.githubusercontent.com/1421719/94219275-75ac0e80-feb4-11ea-9950-b0ad3fbcbb73.png)



The same behavior holds up/down for columns, but it'll prevent you from doing something like... trying to "move_in"  in a way where you're nesting rows in rows or columns in columns... because in that scenario the user should just swap. Again, referring to the graph above, a column with a direct child of a column would be an invalid state.


Ok, another example with four columns, one of which has rows:

![Screenshot from 2020-09-24 21-47-33](https://user-images.githubusercontent.com/1421719/94220086-3979ad80-feb6-11ea-9593-3bbd433c786c.png)

With the 3rd column's tile focused, move_out('down") moves it out of the columns into a row. 

![Screenshot from 2020-09-24 21-48-10](https://user-images.githubusercontent.com/1421719/94220202-7776d180-feb6-11ea-924b-1f6d9f49c091.png)



This introduces a new parent row node above the existing column node and shifts everything from this

![Screenshot from 2020-09-24 22-40-26](https://user-images.githubusercontent.com/1421719/94220478-08e64380-feb7-11ea-8766-cae8a22ea34a.png)


to this

![Screenshot from 2020-09-24 22-40-36](https://user-images.githubusercontent.com/1421719/94220494-100d5180-feb7-11ea-9528-3ae86de91a62.png)


Other things to note:

I have some unit tests and documentation that I'm fleshing out. I am going to go back and add a function for unit tests that would make it so you could setup graphs by passing strings like 'p,p,fl,p,mir,mou' or something and it would do push, push, focus_left, push, move_in(right), move_out(up) to make it easier to setup scenarios for testing

I also want to look into serializing/deserializing the graphs to handle switching on/off workmode or saving to a file to load workspaces later.

OH yeah, there was a bug I encountered with modes where when it's running the code attached to a keybinding it looks through they keybindings and grabs the first one even if it's not part of the current mode. So, if you have multiple modes with a keybinding on the same thing, it'll always use the first mode's keybinding even if it's not active

I fixed it by updating get_keybinding in src/keybindings.rs so that it filters on the current mode which works for the keybindings but NOT for the keybinding to activate/deactivate the mode. That might take some work with rhai so that it treats the activate/deactivate keybinding as also a keybinding under the given mode, but I was having trouble finding where that was determined. I ended up doing this

`KeybindingType::ToggleMode(mode.clone().unwrap_or("none".to_string()));`

and checking if the keybinding's type matches but that seems pretty hacky. I might open a separate issue for this.

Anyway, I have a recent version pushed on my fork but I don't think it's "pull-request ready" especially since it's a big change and also I wanted to know what you thought of the move system I explained above. 

Again, sorry about the dump of text and pictures here.

**Comment by TimUntersberger** [comment]
> I have some unit tests and documentation that I'm fleshing out. I am going to go back and add a function for unit tests that would make it so you could setup graphs by passing strings like 'p,p,fl,p,mir,mou' or something and it would do push, push, focus_left, push, move_in(right), move_out(up) to make it easier to setup scenarios for testing

This seems to be VERY useful.

I also noticed that our naming of keybindings are unintuitive, but I should probably create a separate issue for changing some of the names.

> Again, sorry about the dump of text and pictures here.

I don't mind these massive dumps. I actually like to learn theses dumps as I am always learning something.

> Anyway, I have a recent version pushed on my fork but I don't think it's "pull-request ready" especially since it's a big change and also I wanted to know what you thought of the move system I explained above.

I'll look at your code later this weekend, but for now you can actually open a PR with a `[WIP]` in the beginning until you are ready.

Thank you, I really appreciate your help!

**Comment by ramirezmike** [comment]
I made a WIP pull request. Again, I think it's still got some work left both from a functionality perspective and from a code cleanup perspective. My approach was to make a secondary tile_grid.rs (called "tile_manager.rs" and replace pieces of the original one-by-one until I was able to delete the tile_gris.rs just to keep track of functionality. I'm not married to the names and I think for a final pull request it might be easier if I rename everything back to tile_grid just to remove that noise from the diffs. Also, I figure I'll try to keep it compatible with the OS changes you're making too if it makes more sense to end up in that branch? That shouldn't affect this too much, I think, at least not in a big design change way

The main chunk of code is in tile_manager.rs and it handles all of what I'd consider are nog-specific tile logic. I also made a "graph_wrapper" which is a layer on top of the petgraph stuff to just keep all petgraph references in one place and abstract away some of the graph concepts. Everything else is just updates to the code to use tile_manager functions instead of tile_grid functions. 

A couple other notes
- I did do the tests thing I mentioned above, you can see it in the perform_actions function in tests.rs.
- petgraph has a serialization feature, I think this might be useful for something like... every time you make a change it can serialize and store the current state in a file so that if nog crashes it can reload the last state. Or that can be used for the workmode on/off
- I had a thought about implementing the fullscreen feature as filtering the current graph and storing it in a fullscreen_graph so that the user could like... mark tiles some way to put into fullscreen, like if they wanted to quickly show two tiles side-by-side, and then when they exit it reverts back to the graph they had.
- I have used panic! in a few places just to make it easier if I run into an issue but figure that will be swapped with some logging later

**Comment by TimUntersberger** [comment]
> petgraph has a serialization feature, I think this might be useful for something like... every time you make a change it can serialize and store the current state in a file so that if nog crashes it can reload the last state. Or that can be used for the workmode on/off

I feel like serialization should be done generially on our side with a custom format, to reduce the amout of work you have to do.

>  had a thought about implementing the fullscreen feature as filtering the current graph and storing it in a fullscreen_graph so that the user could like... mark tiles some way to put into fullscreen, like if they wanted to quickly show two tiles side-by-side, and then when they exit it reverts back to the graph they had.

So something like a "scratch workspace"? 

Also lets try to move PR related discussions to the PR.

**Comment by TimUntersberger** [comment]
> I'm not married to the names and I think for a final pull request it might be easier if I rename everything back to tile_grid just to remove that noise from the diffs

My initial idea is/was that the grid has a an "organizer" (or tile manager in this case) and a renderer. The Grid itself has almost no logic and just delegates the work to these two structs.

>  I figure I'll try to keep it compatible with the OS changes you're making too if it makes more sense to end up in that branch

Don't worry about the PR until it is merged.

**Comment by ramirezmike** [comment]
Implemented in #164
